Add ttl querying and locked? method

Martin-Nyaga opened the pull request:
Hey @maltoe, @leandromoreira 

This PR addresses #95. It's a first pass at adding the `get_remaining_ttl` and `locked?` methods and in its current state, should not affect any exisiting behavior at all, apart from loading the new `PTTL` script.

A few comments/questions:
1. I've included some housekeeping changes, I'm happy to remove/move these to another PR depending on your feedback. These are:
- Removed Gemfile.lock and added to gitignore. It was breaking the docker tests since the `ruby:latest` image ships with a newer version of bundler than the one that was used to generate the lockfile. That said, I believe it's not recommended to have Gemfile.lock committed to version control for ruby gems, having well defined constraints in the gemspec (which this gem has) should be enough ([source](https://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/)). This shouldn't cause any material changes, but I'm happy to revert if you disagree.
- Updated the `redis_client` let helper in the tests to use the defined redis host and port, again this was breaking the docker tests as it was defaulting to `localhost`.
2. Presently the code queries all servers for the TTL, and drops the lowest (n/2-1) and picks the lowest TTL from the remaining quorum. I did it this way because it was easier to write as a first pass, but I'm happy to update this if we decide differently in the other conversation.
3. Presently `try_get_remaining_ttl` returns `nil` if either the lock was not found on any servers, or if quorum was not achieved. This feels a little overloaded and there's a chance it could be misleading to the caller. How do you think we should handle the latter case?
4. The tests need to be extended. Off the top of my head, I need to at least assert the behaviour of `get_remaining_ttl` if there was no lock. There may be more test cases that I have missed, but I will review again tomorrow.
5. Note to self: Does the `evalsha` return a string or an integer for the TTL? I will check tomorrow, if it's a string, then this should be cast and the current implementation is incorrect.
6. There is some duplication in `try_get_remaining_ttl`, it wasn't immediately clear how to dry it up, but I'll take another look tomorrow at opportunities for any low-hanging refactoring.

Otherwise, I'm open to feedback and suggestions you may have!

maltoe commented on lib/redlock/client.rb:
```diff
@@ -102,6 +102,21 @@ def lock!(resource, *args)
+    def locked?(lock_info)
+      ttl = get_remaining_ttl(lock_info)
+      !(ttl.nil? || ttl.zero?)
+    end
```
Now that I see it in code: Do you think adding this method is "safe" for users? I mean, you're excluding `ttl == 0`, but what if `ttl == 1`? `locked?` will return `true` and some innocent user of the library will happily continue with their critical code section, even though the lock will be released basically the next millisecond.

In other words: How about we don't add this `locked?` function at all, forcing users to think about their "minimum" remaining ttl to let their jobs continue?

Martin-Nyaga replied:
@maltoe that's a great observation. I'm thinking about my use case, where sometimes, for non-critical reasons like displaying an indicator back to the user, I do indeed just want to check if a resource is locked, and am not very worried if it will unlock immediately after. Here `locked?` is still useful. Perhaps this is something that could be addressed in the documentation instead to make it clear that `locked?` could behave this way, and if you need safety, you should use `get_remaining_ttl`. But we could also delegate this responsibility to the caller. What do you think?

maltoe commented on lib/redlock/client.rb:
```diff
@@ -159,11 +180,20 @@ def unlock(resource, val)
 
+      def get_remaining_ttl(resource, val)
+        recover_from_script_flush do
+          @redis.with { |conn| conn.evalsha @pttl_script_sha, keys: [resource], argv: [val] }
```
this should return an integer

maltoe commented on lib/redlock/client.rb:
```diff
@@ -228,6 +258,22 @@ def lock_instances(resource, ttl, options)
+        # TODO: This nil is overloaded, it could mean that we did not find the
+        # value, or that we could not achieve quorum. Maybe we want to do
+        # something different in the latter case?
+        nil
```
IMO this is ok, and certainly so as the interface of the `get_remaining_ttl()` function. 

> it could mean that we did not find the value, or that we could not achieve quorum

isn't this the same case? at least from the user's perspective?

Martin-Nyaga replied:
Hmm, I guess I'm worried about a situation where we could only reach say 2/5 servers, and so we really don't know the TTL. Now that I think about it, assuming the lock has expired should be fine behaviour from the users perspective,

maltoe commented on lib/redlock/client.rb:
```diff
@@ -228,6 +258,22 @@ def lock_instances(resource, ttl, options)
+      end
+
+      if ttls.size >= @quorum
+        min_ttl = ttls.sort.last(@quorum).min
```
ok, I thought this would have a heavier footprint in the code than `.last(@quorum)` 😄 

Though, I'm now relatively confident that a "reduce" solution and `break` in the middle is in fact better. If we consider the "average" case to be when each server shows almost identical & constant roundtrip time, the `ttls` list will be ordered descending (`[50, 40, 30, 20, 10]`), so we'll usually ditch the values acquired last (and subtract a greater `time_elapsed` from the earlier values, so we don't gain anything).

Sorry for making you take this detour 😊

maltoe commented on spec/client_spec.rb:
```diff
@@ -367,6 +367,36 @@ def redis.with
+      remaining_ttl = lock_manager.get_remaining_ttl(@lock_info)
+      expect(remaining_ttl).to be_within(300).of(ttl)
+    end
+  end
```
* returns `nil` when lock is not held
* test that it returns minimum somehow, probably by mocking redis away

maltoe reviewed (commented):
👍  nice work already

maltoe commented:
> Removed Gemfile.lock

fine with me

maltoe commented on lib/redlock/client.rb:
```diff
@@ -228,6 +258,22 @@ def lock_instances(resource, ttl, options)
 
+    def try_get_remaining_ttl(resource, value)
+      ttls, time_elapsed = timed do
+        @servers.map { |s| s.get_remaining_ttl resource, value }.compact
```
something like

```ruby
@servers.lazy.map { |s| s.get_remaining_ttl resource, value }.select(&:itself).take(@quorum).to_a
```

should work, right?

Martin-Nyaga replied:
Yeah, this looks good 👍

Martin-Nyaga commented:
@maltoe thinking more about the potential use cases (mine included), perhaps we should also allow calling the new methods with just a resource name. Sometimes some other independent non-critical code (which does not hold the lock), would like to know if the lock is locked (by anyone) and do something in response.

So the semantics would be:
- If a lock_info hash is passed, get_remaining_ttl will assert that the lock is held by the caller and return the ttl, This is useful in critical code currently holding the lock.
- If a string is passed, get_remaining_ttl will not make this assertion and return the ttl of whoever is holding the lock. This is useful in non-critical code which does not hold the lock but would like to know it's status.

What do you think?

maltoe commented:
@Martin-Nyaga 

> If a string is passed, get_remaining_ttl will not make this assertion and return the ttl of whoever is holding the lock. This is useful in non-critical code which does not hold the lock but would like to know its status.

Keep in mind that you can have multiple different values on the servers, so you would need to find most common value, see that it still has the quorum, and return the remaining ttl. Also you could then easily implement the `lock_info` case in terms of this "generic" function.

I wonder about overloading (& type checking) the argument though, feels a bit unidiomatic to me. Any other suggestions? Keyword arguments?

Martin-Nyaga commented:
> you would need to find most common value, see that it still has the quorum ... Also you could then easily implement the lock_info case in terms of this "generic" function.

Noted. I'll take this into account.

> I wonder about overloading (& type checking) the argument though, feels a bit unidiomatic to me. Any other suggestions?

Agreed, it could be better. How about something like this:

```ruby
lock_manager.get_remaining_ttl(resource_name) # Returns to any caller
lock_manager.get_remaining_ttl(resource_name, lock_info: lock_info) # Returns to holder of the lock if the lock is valid
```

maltoe commented:
How about 

```ruby
lock_manager.get_remaining_ttl_for_resource(resource_name)
lock_manager.get_remaining_ttl_for_lock(lock_info)
```

Martin-Nyaga commented:
Sure, even better 👍

Martin-Nyaga commented:
@maltoe, I've updated the PR with all the additions from our conversations. The new surface area introduced to the API consists of 4 methods:
```ruby
lock_manager.locked?(resource)
lock_manager.valid_lock?(lock_info)
lock_manager.get_remaining_ttl_for_resource(resource)
lock_manager.get_remaining_ttl_for_lock(lock_info)
```

These are implemented based on a single `try_get_remaining_ttl(resource)`. Tests and documentation are be up to date as well.

Martin-Nyaga commented on lib/redlock/client.rb:
```diff
@@ -228,6 +281,36 @@ def lock_instances(resource, ttl, options)
       end
     end
 
+    def try_get_remaining_ttl(resource)
```
@maltoe ultimately, this has introduced more complexity than anticipated, but it should cover all the bases. Let me know what you think.

Martin-Nyaga commented on spec/client_spec.rb:
```diff
@@ -367,6 +367,111 @@ def redis.with
+        let (:redlock) { Redlock::Client.new(servers) }
+        after(:each) { redlock.unlock(@lock_info) if @lock_info }
+
+        it 'returns the minimum ttl value' do
```
@maltoe there's also a little bit of gymnastics going on here to test this, but it should be correct. Let me know what you think.

maltoe replied:
🤔  yes, would be nicer if the "calculation" part of it (so everything besides the actual redis call) was a pure public function on some module... but again, I think if you start refactoring things like this, it'll be very difficult to stop at some point 😄 

would it be difficult to add a third server that responds with `15_000` and then see that this is returned (for the `.last(@quorum).min` line)?

maltoe commented on lib/redlock/client.rb:
```diff
@@ -102,6 +102,44 @@ def lock!(resource, *args)
+    # currently holds the lock and it has not expired, otherwise the method
+    # returns nil.
+    # Params:
+    # +resource+:: the name of the resource (string) for which to check the ttl
```
☝️  no such param

maltoe commented on lib/redlock/client.rb:
```diff
@@ -122,6 +160,12 @@ class RedisInstance
       eos
 
+      PTTL_SCRIPT = <<-eos
+        if redis.call("exists", KEYS[1]) then
```
i assume you intentionally put this in, but still have to ask: Is the `exists` before `get` actually needed? Doesn't `get` simply return `nil` if the key isn't present?

maltoe replied:
```
$ redis-cli
127.0.0.1:6379> get 'foo'
(nil)
127.0.0.1:6379> pttl 'foo'
(integer) -2
```

Martin-Nyaga replied:
You're right, the code works just as well without it. After testing a bit, it probably wasn't even doing what I thought.

maltoe commented on lib/redlock/client.rb:
```diff
@@ -228,6 +281,37 @@ def lock_instances(resource, ttl, options)
+        # Return the  minimum TTL of an N/2+1 selection. It will always be
+        # correct (it will guarantee that at least N/2+1 servers have a TTL that
+        # value or longer)
+        min_ttl = ttls.sort.last(@quorum).min
```
technically this is the same as `ttls.sort.last(@quorum).first`, isn't it? saves you one list scan to get the `min`

Martin-Nyaga commented:
@maltoe I've addressed all the comments from your review.

At the moment, The tests are still failing because `Hash#transform_values` is only available in 2.4+. This would be fixed by #90.

RE: Refactoring the minimum TTL test, I took a quick stab at some refactoring, but I decided against it - it was a bit of a rabbit hole as you mentioned. Perhaps it's better addressed in a separate PR addressing some general refactoring of the test suite. I can see how a `Redlock::TestHelper` module (or something like that) can provide helper methods to alleviate duplication/complexity in the present suite. When I have some time, I'm happy to give this a go as well if you think it's a good idea.

leandromoreira commented:
can I merge?

maltoe commented:
@leandromoreira it has a conflict in the `Gemfile.lock` and it would be nice to see it green with the new CI ruby versions (merged #90 earlier)

any idea why Travis seems to do nothing on #97 ?

Martin-Nyaga commented:
@maltoe I'm happy to rebase on master, shortly.

leandromoreira commented:
> @leandromoreira it has a conflict in the `Gemfile.lock` and it would be nice to see it green with the new CI ruby versions (merged #90 earlier)
> 
> any idea why Travis seems to do nothing on #97 ?

no idea, it's waiting, let's see if takes more time I think we can retry

Martin-Nyaga commented:
> it has a conflict in the `Gemfile.lock`
Rebased, conflicts solved 👍

leandromoreira commented:
I think travis is facing some issues :/

it doesn't leave the received status

<img width="1082" alt="image" src="https://user-images.githubusercontent.com/55913/98991643-e6fa5c00-250a-11eb-8bfa-e7a054c18a5f.png">

leandromoreira commented:
Can we merge this? :)

leandromoreira commented:
@Martin-Nyaga thanks a lot for the contribution and sorry for the delay =/

Martin-Nyaga commented:
Cheers @leandromoreira, no worries. 👍
